docs/GDPs/gdp-0002-regions.rst

```diff
@@ -0,0 +1,412 @@
+Each endpoint is itself a pair of a `gt4py.definitions.Interval` and an integer offset.
+For example, to fill a strip the `west` strip of a I-J plane of size 3, the region specification would be:
+
+.. code-block:: python
+
+    (((Interval.START, 0), (Interval.START, 2)), ((Interval.START, 0), (Interval.STOP, 0)))
+
+This specifies an iteration region over the first 3 points in ``i``` (``0,1,2``) from the application domain edge, and all ``j`` in the grid.
+Again, the computation will only happen on portions of the region that have an intersection with the written-to field's extended compute domain.
+
+
+Additional run-time information
++++++++++++++++++++++++++++++++
+
+GT4Py currently stores information about the offsets between fields and the compute domain by specifying origins for each field that are referred to by ``(0,0)`` in the compute domain.
+Therefore the additional information that is needed for regional computation is to specify (1) the total size of the application domain and (2) how the fields relate to the application domain.
+
+The total size of the application domain can be represented as a numeric tuple of two `ints`.
+
+Currently, GT4Py ties the compute domain origin to each field's origin when a single origin is passed at stencil run time.
+**Specifying an origin to a stencil call overwrites the origin for each field.**
+When the storage rewrite is complete, fields will also have halo metadata, so this can be used to align the non-halo portion of a field relative to the application domain.
+The easiest method is to specify the origin offset between the bottom left non-halo point in the field (the natural/default origin) to the application domain.
+Much of the time this will be ``(0, 0)`` (in which case it does not need to be specified), but could be non-zero for staggered fields.
+This information should be collected in a dictionary of field name to numeric tuple, and passed to stencils at execution time.
+
+The call could then look like:
+
+.. code-block:: python
+
+    stencil(uc, vc, cosa, rsina, ut, ub, dt5, dt4,
+        app_domain=(grid.ni, grid.nj), app_origins={ub: (1,0)})
+
+
+Temporary fields
+^^^^^^^^^^^^^^^^
+
+The information above is sufficient for field arguments to the stencil, but what about temporaries? Assigning to temporaries in a region should be allowed, so GT4Py needs to generate an offset automatically from the application domain for these fields. The natural way to determine this is to have the `origin` of the temporary field be the origin of the compute domain.
+
+Take for example the double Laplacian example, with a special case before applying the second Laplacian:
+
+.. code-block:: python
+
+    @gtscript.stencil()
+    def double_lap(in_f: gtscript.Field[float],
+                out_f: gtscript.Field[float]):
+        with computation(PARALLEL), interval(...):
+            tmp_f = lap(in_f)
+            with region(LeftEdge(-1, 0)):
+                tmp_f = 1.0
```

Comment:
```suggestion
                tmp_f = in_f[-1,0,0]
```
Would that be allowed? Why not?

If I replace `tmp_f` by an external field and remove the second lap, will the region be executed or not (i.e. will the external field have 1.0 in the second halo line.

I need to read it more carefully tomorrow, to see if I can find the answer myself.

Reply:
That should still work, yes. It is accessing the last halo line for `in_f` in that case. It would extend the read extent for `in_f`.

Reply:
Right, now let's trigger a read from one line more (which could just happen by merging another unrelated operation into this stencil). So 3 halo lines which are all consumed. What happens at the second halo line?

Reply:
Well, that wouldn't be valid anyway (there is no data there in `in_f`). You would have to shrink the computational domain for the stencil. If you didn't specify the domain it would automatically shrink it to `domain=(2,2)`, and the stencil would not error.